lib/winapi/timezoneapi.c

```diff
@@ -0,0 +1,197 @@
+        return TIME_ZONE_ID_INVALID;
+    }
+    queryStatus = ExQueryNonVolatileSetting(XC_TZ_STD_NAME, &type, standardName, sizeof(standardName), NULL);
+    if (!NT_SUCCESS(queryStatus)) {
+        SetLastError(RtlNtStatusToDosError(queryStatus));
+        return TIME_ZONE_ID_INVALID;
+    }
+
+    if (dayLightEndDate.Month > 12 || dayLightStartDate.Month > 12 ||
+        dayLightEndDate.DayOfWeek > 6 || dayLightStartDate.DayOfWeek > 6 ||
+        dayLightEndDate.Hour > 23 || dayLightStartDate.Hour > 23) {
+        return TIME_ZONE_ID_INVALID;
+    }
+
+    // Xbox stores up to 4 characters here
+    for (UCHAR i = 0; i < 4; i++) {
+        lpTimeZoneInformation->StandardName[i] = standardName[i];
+        lpTimeZoneInformation->DaylightName[i] = dayLightName[i];
+    }
+
+    // There was no DST info
+    if (dayLightEndDate.Month == 0 || dayLightStartDate.Month == 0) {
+        return TIME_ZONE_ID_UNKNOWN;
+    }
+
+    lpTimeZoneInformation->StandardDate.wMonth = dayLightEndDate.Month;
+    lpTimeZoneInformation->StandardDate.wDayOfWeek = dayLightEndDate.DayOfWeek;
+    lpTimeZoneInformation->StandardDate.wDay = dayLightEndDate.Day;
+    lpTimeZoneInformation->StandardDate.wHour = dayLightEndDate.Hour;
+    lpTimeZoneInformation->StandardBias = standardBias;
+
+    lpTimeZoneInformation->DaylightDate.wMonth = dayLightStartDate.Month;
+    lpTimeZoneInformation->DaylightDate.wDayOfWeek = dayLightStartDate.DayOfWeek;
+    lpTimeZoneInformation->DaylightDate.wDay = dayLightStartDate.Day;
+    lpTimeZoneInformation->DaylightDate.wHour = dayLightStartDate.Hour;
+    lpTimeZoneInformation->DaylightBias = dayLightBias;
+
+    // Now we need to determine if the current time is within the DST range
+    KeQuerySystemTime(&kTime);
+    RtlTimeToTimeFields(&kTime, &dateNow);
+
+    // Determine what day in the month DST starts and ends
+    UCHAR dayLightStartDay = GetActualDay(dateNow.Year, &dayLightStartDate);
+    UCHAR dayLightEndDay = GetActualDay(dateNow.Year, &dayLightEndDate);
+
+    // Fix month wrapping. i.e if DST is from Oct(10) to Feb(2) and the current month is Jan(1) we should be in DST
+    // Applying a simple check would result is 10 < 1 < 2 = false and incorrect
+    // To account for this, we offset the months in the following year by 12
+    // This results in the comparison 10 < 13(12+1) < 14(12+2) = true and is correct
+    if (dayLightEndDate.Month < dayLightStartDate.Month)
+    {
+        if (dateNow.Month <= dayLightEndDate.Month) {
+            dateNow.Month += 12;
+        }
+        dayLightEndDate.Month += 12;
+    }
+
+    // Now that month wrapping is fixed, we convert the dates to minutes from the start of the year corresponding to dayLightStartDate
+    // The conversion to minutes allows for easier comparison. The minute resolution allows for timezones with 0.5 hour offsets.
+    const LONG MPD = 24 * 60;
+    const LONG MPM = 31 * MPD; //  31 is ok providing it's consistent
```

Comment:
nit: upper-case variable names should probably be reserved for macros.

It's also not clear what MPD and MPM stand for (although it's probably minutes-per-day and minutes-per-month)

Reply:
made lower case. Yes minutes per day and per month